File: docs/code/merge_sort.ts
/**
 * 归并排序 (Merge Sort) 实现
 * 时间复杂度: O(n log n)
 * 空间复杂度: O(n)
 */

/**
 * 归并排序主函数
 * 采用自顶向下的递归方式实现
 */
export function mergeSort(arr: number[]): number[] {
  if (arr.length <= 1) {
    return arr;
  }

  // 分割阶段: 将数组分成两半
  const mid = Math.floor(arr.length / 2);
  const leftHalf = arr.slice(0, mid);
  const rightHalf = arr.slice(mid);

  // 递归排序左右两半
  const leftSorted = mergeSort(leftHalf);
  const rightSorted = mergeSort(rightHalf);

  // 合并阶段: 合并两个有序数组
  return merge(leftSorted, rightSorted);
}

/**
 * 合并两个有序数组
 * 使用双指针技术
 */
function merge(left: number[], right: number[]): number[] {
  const result: number[] = [];
  let i = 0;
  let j = 0;

  // 比较两个数组的头部元素，将较小的放入结果
  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]) {
      result.push(left[i]);
      i++;
    } else {
      result.push(right[j]);
      j++;
    }
  }

  // 将剩余元素追加到结果中
  while (i < left.length) {
    result.push(left[i]);
    i++;
  }
  while (j < right.length) {
    result.push(right[j]);
    j++;
  }

  return result;
}

/**
 * 归并排序（返回排序过程中的所有状态）
 * 用于可视化展示
 */
export interface SortStep {
  array: number[];
  activeIndices: number[];
  mergeIndices: number[];
  sortedIndices: number[];
  description: string;
}

export function mergeSortWithSteps(originalArr: number[]): SortStep[] {
  const steps: SortStep[] = [];
  const arr = [...originalArr];

  // 记录初始状态
  steps.push({
    array: [...arr],
    activeIndices: [],
    mergeIndices: [],
    sortedIndices: [],
    description: '初始数组',
  });

  function mergeSortRecursive(left: number, right: number): void {
    if (left >= right) {
      if (left === right) {
        steps.push({
          array: [...arr],
          activeIndices: [],
          mergeIndices: [],
          sortedIndices: [left],
          description: `单个元素 arr[${left}] = ${arr[left]}`,
        });
      }
      return;
    }

    const mid = Math.floor((left + right) / 2);

    mergeSortRecursive(left, mid);
    mergeSortRecursive(mid + 1, right);

    mergeInPlace(left, mid, right);
  }

  function mergeInPlace(left: number, mid: number, right: number): void {
    const leftArr = arr.slice(left, mid + 1);
    const rightArr = arr.slice(mid + 1, right + 1);

    let i = 0;
    let j = 0;
    let k = left;

    while (i < leftArr.length && j < rightArr.length) {
      steps.push({
        array: [...arr],
        activeIndices: [left + i, mid + 1 + j],
        mergeIndices: Array.from({ length: k - left }, (_, idx) => left + idx),
        sortedIndices: [],
        description: `比较 ${leftArr[i]} 和 ${rightArr[j]}`,
      });

      if (leftArr[i] <= rightArr[j]) {
        arr[k] = leftArr[i];
        i++;
      } else {
        arr[k] = rightArr[j];
        j++;
      }
      k++;
    }

    while (i < leftArr.length) {
      arr[k] = leftArr[i];
      i++;
      k++;
    }

    while (j < rightArr.length) {
      arr[k] = rightArr[j];
      j++;
      k++;
    }

    // 标记已排序的范围
    const sortedIndices: number[] = [];
    for (let idx = left; idx <= right; idx++) {
      sortedIndices.push(idx);
    }

    steps.push({
      array: [...arr],
      activeIndices: [],
      mergeIndices: [],
      sortedIndices,
      description: `合并区间 [${left}..${right}]`,
    });
  }

  mergeSortRecursive(0, arr.length - 1);

  // 最终状态
  steps.push({
    array: [...arr],
    activeIndices: [],
    mergeIndices: [],
    sortedIndices: arr.map((_, i) => i),
    description: '排序完成! 🎉',
  });

  return steps;
}

// ============ 测试代码 ============
function test() {
  const testCases = [
    [38, 27, 43, 3, 9, 82, 10],
    [5, 2, 8, 1, 9],
    [1],
    [],
    [1, 2, 3, 4, 5],
    [5, 4, 3, 2, 1],
  ];

  console.log('=== 归并排序测试 ===');
  for (const arr of testCases) {
    const original = [...arr];
    const sorted = mergeSort(arr);
    console.log(`原始: ${JSON.stringify(original)} -> 排序后: ${JSON.stringify(sorted)}`);
  }

  console.log('\n=== 归并排序步骤（前 10 步） ===');
  const steps = mergeSortWithSteps([38, 27, 43, 3, 9, 82, 10]);
  for (let i = 0; i < Math.min(10, steps.length); i++) {
    const s = steps[i];
    console.log(`Step ${i}: ${s.description} => [${s.array}]`);
  }
  console.log(`... 共 ${steps.length} 步`);
}

test();

File: docs/code/merge_sort.py
"""
归并排序 (Merge Sort) 实现
时间复杂度: O(n log n)
空间复杂度: O(n)
"""

def merge_sort(arr: list[int]) -> list[int]:
    """
    归并排序主函数
    采用自顶向下的递归方式实现
    """
    if len(arr) <= 1:
        return arr

    # 分割阶段: 将数组分成两半
    mid = len(arr) // 2
    left_half = arr[:mid]
    right_half = arr[mid:]

    # 递归排序左右两半
    left_sorted = merge_sort(left_half)
    right_sorted = merge_sort(right_half)

    # 合并阶段: 合并两个有序数组
    return merge(left_sorted, right_sorted)


def merge(left: list[int], right: list[int]) -> list[int]:
    """
    合并两个有序数组
    使用双指针技术
    """
    result = []
    i = j = 0

    # 比较两个数组的头部元素，将较小的放入结果
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # 将剩余元素追加到结果中
    result.extend(left[i:])
    result.extend(right[j:])

    return result


def merge_sort_in_place(arr: list[int], left: int = 0, right: int | None = None) -> None:
    """
    原地归并排序 (优化空间版本)
    直接修改原数组，使用临时数组辅助合并
    """
    if right is None:
        right = len(arr) - 1

    if left < right:
        mid = (left + right) // 2

        merge_sort_in_place(arr, left, mid)
        merge_sort_in_place(arr, mid + 1, right)

        merge_in_place(arr, left, mid, right)


def merge_in_place(arr: list[int], left: int, mid: int, right: int) -> None:
    """
    原地合并两个有序子数组 [left..mid] 和 [mid+1..right]
    """
    # 创建临时数组
    left_arr = arr[left:mid + 1]
    right_arr = arr[mid + 1:right + 1]

    i = j = 0
    k = left

    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1

    while i < len(left_arr):
        arr[k] = left_arr[i]
        i += 1
        k += 1

    while j < len(right_arr):
        arr[k] = right_arr[j]
        j += 1
        k += 1


# ============ 测试代码 ============
if __name__ == "__main__":
    test_cases = [
        [38, 27, 43, 3, 9, 82, 10],
        [5, 2, 8, 1, 9],
        [1],
        [],
        [1, 2, 3, 4, 5],
        [5, 4, 3, 2, 1],
    ]

    print("=== 归并排序测试 ===")
    for arr in test_cases:
        original = arr.copy()
        sorted_arr = merge_sort(arr)
        print(f"原始: {original} -> 排序后: {sorted_arr}")

    # 测试原地排序
    print("\n=== 原地归并排序测试 ===")
    for arr in test_cases:
        original = arr.copy()
        merge_sort_in_place(arr)
        print(f"原始: {original} -> 排序后: {arr}")
